contributing.md

```diff
@@ -0,0 +1,276 @@
+to indicate when and how they have been addressed.
+
+If your branch has become out of sync with *upstream* then conflicts may
+arise. Sometimes these cannot be automatically resolved and you will
+need to resolve them by hand. Github provides instructions for this, or
+you can follow this routine,
+
+``` sh
+# add upstream as a remote if you have not already
+$ git remote add upstream git@github.com:stfc/janus-core.git
+# get the changes to upstream since you started working on your issue
+$ git fetch upstream
+# merge these changes into your branch (assuming you want to merge into the main branch on upstream)
+$ git merge upstream/main
+# resolve any conflicts
+# push to your fork
+$ git push
+```
+
+Alternatively you may use rebase which will replay the changes you made
+in your branch on top of *upstream/main* however be sure you understand
+the differences between merge and rebase
+
+``` sh
+# add upstream as a remote if you have not already
+$ git remote add upstream git@github.com:stfc/janus-core.git
+# get the changes to upstream since you started working on your issue
+$ git fetch upstream
+# merge these changes into your branch (assuming you want to merge into the main branch on upstream)
+$ git rebase -i upstream/main
+# resolve any conflicts
+# push to your fork
+$ git push
+```
+
+
+### Advanced git
+
+#### Keeping your fork in sync with project
+
+By adding two remotes, one for *upstream* and one for your *fork* it is
+possible to keep your *fork* in sync with *upstream*. This will greatly
+simplify merge requests. Github also offers a sync functionality in their
+web ui that achieves the same.
+
+``` sh
+# clone your fork
+$ git clone git@github.com:username/janus-core.git janus-core-username
+pushd janus-core-username
+# add a remote for upstream
+$ git remote add upstream git@github.com:stfc/janus-core.git
+```
+
+These commands need to be done only once. `git remote -v` shall show you
+the origin and project fetch and push links
+
+``` sh
+$ git remote -v
+origin  git@github.com:username/janus-core.git (fetch)
+origin  git@github.com:username/janus-core.git (push)
+upstream git@github.com:stfc/janus-core.git (fetch)
+upstream git@github.com:stfc/janus-core.git (push)
+```
+
+When you need to sync your *fork* with *upstream*, do the following,
+
+``` sh
+# get the latest commits from upstream
+$ git fetch upstream
+# ensure you are in the main branch of your fork
+$ git checkout main
+# merge your main branch into the main branch of upstream
+$ git merge upstream/main
+# push these changes back to the remote of your fork (origin)
+$ git push
+```
+
+of course one can use a similar process to merge any other branch or
+available projects.
+
+#### Rebasing commits
+
+When working on an issue you may use multiple commits. When you are
+ready to create a merge request, you should squash your changes into one
+commit in order to keep *upstream* clean. This is most easily achieved with
+an interactive rebase.
```

Comment:
Do we want to recommend this? Before/during review, my preference is generally squash minor commits, but keep some separation of more substantial changes, which I find makes it easier to review, as long as we squash down on merging with main.

Reply:
Also "Squash and Merge" is a perfectly valid way of doing it.

Reply:
I suggest it mainly to force people to have modular reviews with one and only one feature in. I know can be overhead, but once you need to debug pays off.